src/dock.js

```diff
@@ -0,0 +1,544 @@
+      this.draggedOut()
+    }
+  }
+
+  handleDragEnd () {
+    this.draggedOut()
+  }
+
+  draggedOut () {
+    this.setState({showDropTarget: false})
+    window.removeEventListener('drag', this.handleDrag)
+    window.removeEventListener('dragend', this.handleDragEnd)
+  }
+
+  // Determine whether the cursor is within the dock hover area. This isn't as simple as just using
+  // mouseenter/leave because we want to be a little more forgiving. For example, if the cursor is
+  // over the footer, we want to show the bottom dock's toggle button.
+  pointWithinHoverArea (point, includeButtonWidth) {
+    const dockBounds = this.innerElement.getBoundingClientRect()
+    // Copy the bounds object since we can't mutate it.
+    const bounds = {
+      top: dockBounds.top,
+      right: dockBounds.right,
+      bottom: dockBounds.bottom,
+      left: dockBounds.left
+    }
+
+    // Include all panels that are closer to the edge than the dock in our calculations.
+    switch (this.location) {
+      case 'right':
+        bounds.right = Number.POSITIVE_INFINITY
+        break
+      case 'bottom':
+        bounds.bottom = Number.POSITIVE_INFINITY
+        break
+      case 'left':
+        bounds.left = 0
+        break
+    }
+
+    // The area used when detecting "leave" events is actually larger than when detecting entrances.
+    if (includeButtonWidth) {
+      const affordance = 20
+      const toggleButtonSize = 50 / 2 // This needs to match the value in the CSS.
+      switch (this.location) {
+        case 'right':
+          bounds.left -= toggleButtonSize + affordance
+          break
+        case 'bottom':
+          bounds.top -= toggleButtonSize + affordance
+          break
+        case 'left':
+          bounds.right += toggleButtonSize + affordance
+          break
+      }
+    }
+    return rectContainsPoint(bounds, point)
+  }
+
+  getInitialSize () {
+    let initialSize
+    // The item may not have been activated yet. If that's the case, just use the first item.
+    const activePaneItem = this.paneContainer.getActivePaneItem() || this.paneContainer.getPaneItems()[0]
+    if (activePaneItem != null) {
+      initialSize = getPreferredInitialSize(activePaneItem, this.location)
+    }
+    return initialSize == null ? DEFAULT_INITIAL_SIZE : initialSize
+  }
+
+  serialize () {
+    return {
+      deserializer: 'Dock',
+      size: this.state.size,
+      paneContainer: this.paneContainer.serialize(),
+      open: this.state.open
+    }
+  }
+
+  deserialize (serialized, deserializerManager) {
+    this.paneContainer.deserialize(serialized.paneContainer, deserializerManager)
+    this.setState({
+      size: serialized.size,
+      // If no items could be deserialized, we don't want to show the dock (even if it was open last time)
+      open: serialized.open && (this.paneContainer.getPaneItems().length > 0)
+    })
+  }
+
+  // PaneContainer-delegating methods
+
+  getPanes () {
```

Comment:
Can you add documentation for these `PaneContainer`-delegating methods, making them public? Docks should have a similar set of public methods as `WorkspaceCenter`.